pkg/meta/distributed_lock.go

```diff
@@ -0,0 +1,314 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package meta
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/pkg/parser/model"
+	"github.com/pingcap/tidb/pkg/util/intest"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	"go.uber.org/zap"
+)
+
+// DataStore is an interface to create transaction.
+type DataStore interface {
+	Begin(ctx context.Context) (txn DataTxn, err error)
+}
+
+// DataTxn is an interface to operate data in transaction.
+type DataTxn interface {
+	Rollback()
+	Commit() error
+
+	StartTS() uint64
+	Set(key, value []byte) error
+	Get(key []byte) ([]byte, error)
+	Del(key []byte) error
+}
+
+type lockContent struct {
+	StartTS    uint64 `json:"start_ts"`
+	InstanceID string `json:"instance_id"`
+}
+
+func newLockContent(startTS uint64, instanceID string) lockContent {
+	return lockContent{
+		StartTS:    startTS,
+		InstanceID: instanceID,
+	}
+}
+
+func (c lockContent) encode() ([]byte, error) {
+	return json.Marshal(c)
+}
+
+func (c *lockContent) decode(data []byte) error {
+	return json.Unmarshal(data, c)
+}
+
+// DistributedLock is a distributed lock implementation based on meta key transaction.
```

Comment:
Can you explain how do we implement lease timeout behaviour? For example, every clients check xx TS to know it's timed out and try to acquire the lock, and contention will be found when KV transaction commits?